File: src/main/java/org/railwaystations/rsapi/core/ports/in/ManageInboxUseCase.java
package org.railwaystations.rsapi.core.ports.in;

import org.railwaystations.rsapi.core.model.InboxCommand;
import org.railwaystations.rsapi.core.model.InboxEntry;
import org.railwaystations.rsapi.core.model.InboxResponse;
import org.railwaystations.rsapi.core.model.InboxStateQuery;
import org.railwaystations.rsapi.core.model.ProblemReport;
import org.railwaystations.rsapi.core.model.PublicInboxEntry;
import org.railwaystations.rsapi.core.model.User;

import javax.validation.constraints.NotNull;
import java.io.InputStream;
import java.util.List;

public interface ManageInboxUseCase {

    InboxResponse reportProblem(ProblemReport problemReport, User user, String clientInfo);

    List<PublicInboxEntry> publicInbox();

    List<InboxStateQuery> userInbox(@NotNull User user, List<Long> ids);

    List<InboxEntry> listAdminInbox(@NotNull User user);

    long countPendingInboxEntries();

    String getNextZ();

    InboxResponse uploadPhoto(String clientInfo, InputStream body, String stationId,
                              String country, String contentType, String stationTitle,
                              Double latitude, Double longitude, String comment,
                              boolean active, User user);

    void changeStationTitle(InboxCommand command);

    void rejectInboxEntry(InboxCommand command);

    void importPhoto(InboxCommand command);

    void importMissingStation(InboxCommand command);

    void updateStationActiveState(InboxCommand command, boolean active);

    void deleteStation(InboxCommand command);

    void deletePrimaryPhoto(InboxCommand command);

    void markProblemReportSolved(InboxCommand command);

    void updateLocation(InboxCommand command);

    void markPrimaryPhotoOutdated(InboxCommand command);
}

File: src/main/kotlin/org/railwaystations/rsapi/core/ports/ManageInboxUseCase.kt
package org.railwaystations.rsapi.core.ports

import org.railwaystations.rsapi.core.model.InboxCommand
import org.railwaystations.rsapi.core.model.InboxEntry
import org.railwaystations.rsapi.core.model.InboxResponse
import org.railwaystations.rsapi.core.model.InboxStateQuery
import org.railwaystations.rsapi.core.model.ProblemReport
import org.railwaystations.rsapi.core.model.PublicInboxEntry
import org.railwaystations.rsapi.core.model.User
import java.io.InputStream

interface ManageInboxUseCase {
    fun reportProblem(problemReport: ProblemReport, user: User, clientInfo: String?): InboxResponse

    fun publicInbox(): List<PublicInboxEntry>

    fun userInbox(user: User): List<InboxStateQuery>

    fun userInbox(user: User, ids: List<Long>): List<InboxStateQuery>

    fun listAdminInbox(user: User): List<InboxEntry>

    fun countPendingInboxEntries(): Long

    val nextZ: String

    fun uploadPhoto(
        clientInfo: String?, body: InputStream?, stationId: String?,
        countryCode: String?, contentType: String?, stationTitle: String?,
        latitude: Double?, longitude: Double?, comment: String?,
        active: Boolean, user: User
    ): InboxResponse

    fun changeStationTitle(command: InboxCommand)

    fun rejectInboxEntry(command: InboxCommand)

    fun importPhoto(command: InboxCommand)

    fun importMissingStation(command: InboxCommand)

    fun updateStationActiveState(command: InboxCommand, active: Boolean)

    fun deleteStation(command: InboxCommand)

    fun deletePhoto(command: InboxCommand)

    fun markProblemReportSolved(command: InboxCommand)

    fun updateLocation(command: InboxCommand)

    fun markPhotoOutdated(command: InboxCommand)
}
